README.md
# video.js hls.js source handler

Based on https://github.com/streamroot/videojs5-hlsjs-source-handler, it replaces hls.js with [hlsjs-p2p-wrapper](https://github.com/streamroot/hlsjs-p2p-wrapper).

## Testing locally

##### build the project with :
```
$ npm install
$ npm run build
```

##### start a http server at the root of the project
```
$ npm run dev
```

##### open http://localhost:8080/debug.html


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Dev]

## [Unreleased]

## [0.3.1] - 2017-02-24
### Fixed
- Manifest deployment

## [0.3.0] - 2017-02-24
### Changed
- Replace Grunt with Webpack.

## [0.2.20] - 2016-12-08
### Updated
- hlsjs-p2p-bundle version to ^4.0.0

## [0.2.12] - 2016-11-10
### Fixed
- Broken fallback logic when not in autoplay

## [0.2.4] - 2016-10-18
### Fixed
- Broken playback when video source was set in video tag. It was caused by source handler disposal, triggered by video.js. See comments in `lib/videojs5-hlsjs-source-handler.js` for detailed explanation.

### Removed
- Quality picker dependency in `package.json` and initialization in test page. It's usage was removed in modifications, required for the Brightcove plugin.

## [0.2.3] - 2016-09-27
### Changed
- Modifications required to use with Brightcove plugin: Hls.js is initialized lazy, only at point when media is requesting data. This way we can be sure that plugin is initialized already and our player instance carries P2P config information.

### Added
- Publish to NPM

## [0.1.4] - 2016-08-11
### Changed
- PreProd distribution name to videojs5-hlsjs-source-handler

### Fixed
- Starting loading before play (now loads when media element requests data)
[0.2.35]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.2.20...v0.2.35[0.3.0]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.2.20...v0.3.0
[0.3.1]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.3.0...v0.3.1[0.3.2]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.3.1...v0.3.2[0.3.3]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.3.1...v0.3.3[0.3.4]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.3.1...v0.3.4
[0.3.5]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.3.1...v0.3.5
[0.3.6]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.3.1...v0.3.6
[0.3.7]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.3.1...v0.3.7
[0.3.8]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.3.1...v0.3.8
[0.3.9]: https://github.com/streamroot/videojs5-hlsjs-p2p-source-handler/compare/v0.3.1...v0.3.9


Gruntfile.js
module.exports = function(grunt) {
    require("matchdep").filterDev("grunt-*").forEach(grunt.loadNpmTasks);

    grunt.registerTask('default', ['browserify']);

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        browserify: {
            main: {
                src: 'lib/main.js',
                dest: 'debug/videojs5-hlsjs-p2p-source-handler.js',
                options:  {
                    browserifyOptions: {
                        debug: true
                    },
                    watch: true,
                    keepAlive: true
                }
            },
            dist: {
                src: 'lib/main.js',
                dest: 'dist/videojs5-hlsjs-p2p-source-handler.js',
                options:  {
                    browserifyOptions: {
                        debug: false,
                    },
                    watch: false,
                    keepAlive: false,
                }
            }
        },
        uglify: {
            options: {
                mangle: true,
                compress: {
                    drop_console: true
                },
                beautify: false
            },
            dist: {
                files: {
                    'dist/videojs5-hlsjs-p2p-source-handler.min.js': 'dist/videojs5-hlsjs-p2p-source-handler.js'
                }
            }
        }
    });


    grunt.registerTask('build', 'build dist scripts', ['browserify:dist', 'uglify:dist']);


}


lib/vjs-hls.js
require("videojs-quality-picker");

var attachVideojsStreamrootProvider = function (window, videojs, Hls) {
    function StreamrootProviderHLS (source, tech) {
        tech.name_ = 'streamrootHLS';

        var _video = tech.el();
        var _hls;

        function initialize() {
            var hlsjsConfig = tech.options_.hlsjsConfig || {};
            var p2pConfig = tech.options_.p2pConfig || {};

            _hls = new Hls(hlsjsConfig, p2pConfig);
            _hls.on(Hls.Events.ERROR, _onError);
            _hls.on(Hls.Events.MANIFEST_PARSED, _onMetaData);

            _hls.attachMedia(_video);
        }

        this.dispose = function () {
            _hls.destroy();
        };

        function load(source) {
            _hls.loadSource(source.src);
        }

        function switchQuality(qualityId, trackType) {
            _hls.nextLevel = qualityId;
        }

        function _onError(event, data) {
            console.error('HLS.js error: ' + data.type + ' - fatal: ' + data.fatal + ' - ' + data.details);
        }

        function _onMetaData(event, data) {
            var cleanTracklist = [];

            if (data.levels.length > 1) {
                var autoLevel = {
                    id: -1,
                    label: "auto",
                    selected: -1 === _hls.manualLevel
                };
                cleanTracklist.push(autoLevel);
            }

            data.levels.forEach(function(level, index) {
                var quality = {}; // Don't write in level (shared reference with Hls.js)
                quality.id = index;
                quality.selected = index === _hls.manualLevel;
                quality.label = _levelLabel(level);

                cleanTracklist.push(quality);
            });

            var payload = {
                qualityData: {video: cleanTracklist},
                qualitySwitchCallback: switchQuality
            };

            tech.trigger('loadedqualitydata', payload);

            function _levelLabel(level) {
                if (level.height) return level.height + "p";
                else if (level.width) return Math.round(level.width * 9 / 16) + "p";
                else if (level.bitrate) return (level.bitrate / 1000) + "kbps";
                else return 0;
            }
        }

        initialize();
        load(source);
    }

    if (Hls.isSupported()) {
        videojs.getComponent('Html5').registerSourceHandler({

            canHandleSource: function (source) {

                var hlsTypeRE = /^application\/x-mpegURL$/i;
                var hlsExtRE = /\.m3u8/i;
                var result;

                if (hlsTypeRE.test(source.type)) {
                    result = 'probably';
                } else if (hlsExtRE.test(source.src)) {
                    result = 'maybe';
                } else {
                    result = '';
                }

                return result;
            },

            handleSource: function (source, tech) {

                if (tech.hlsProvider) {
                    tech.hlsProvider.dispose();
                }

                tech.hlsProvider = new StreamrootProviderHLS(source, tech);

                return tech.hlsProvider;
            }

        }, 0);

    } else {
        console.error("Hls.js is not supported in this browser!");
    }

    videojs.StreamrootProviderHLS = StreamrootProviderHLS;
};

module.exports = attachVideojsStreamrootProvider;


lib/videojs5-hlsjs-source-handler.js
var assign = require('lodash.assign');

var attachVideojsStreamrootProvider = function (window, videojs, Hls) {

    function StreamrootProviderHLS (source, tech) {
        tech.name_ = 'streamrootHLS';

        var _video = tech.el();
        var _hls;
        var _errorCounts = {};
        var _duration = null;

        _video.addEventListener('error', function(evt) {
            var errorTxt,mediaError=evt.currentTarget.error;

            switch(mediaError.code) {
                case mediaError.MEDIA_ERR_ABORTED:
                    errorTxt = "You aborted the video playback";
                    break;
                case mediaError.MEDIA_ERR_DECODE:
                    errorTxt = "The video playback was aborted due to a corruption problem or because the video used features your browser did not support";
                    _handleMediaError();
                    break;
                case mediaError.MEDIA_ERR_NETWORK:
                    errorTxt = "A network error caused the video download to fail part-way";
                    break;
                case mediaError.MEDIA_ERR_SRC_NOT_SUPPORTED:
                    errorTxt = "The video could not be loaded, either because the server or network failed or because the format is not supported";
                    break;
            }

            console.error("MEDIA_ERROR: ", errorTxt);
        });

        function initialize() {
            _initHlsjs();
        }

        this.duration = function () {
            return _duration || _video.duration || 0;
        };

        // See comment for `initialize` method.
        this.dispose = function () {
            _video.removeEventListener('play', _startLoad);

            _hls.destroy();

            tech.setSource(source);
        };

        function switchQuality(qualityId, trackType) {
            _hls.nextLevel = qualityId;
        }

        function _handleMediaError() {
            if (_errorCounts[Hls.ErrorTypes.MEDIA_ERROR] === 1) {
                console.info("trying to recover media error");
                _hls.recoverMediaError();
            } else if (_errorCounts[Hls.ErrorTypes.MEDIA_ERROR] === 2) {
                console.info("2nd try to recover media error (by swapping audio codec");
                _hls.swapAudioCodec();
                _hls.recoverMediaError();
            } else if (_errorCounts[Hls.ErrorTypes.MEDIA_ERROR] > 2) {
                console.info("bubbling media error up to VIDEOJS");
                error.code = 3;
                tech.error = function() { return error; };
                tech.trigger('error');
            }
        }

        function _onError(event, data) {
            var error = {
                message: ('HLS.js error: ' + data.type + ' - fatal: ' + data.fatal + ' - ' + data.details),
            };
            console.error(error.message);

            // increment/set error count
            _errorCounts[data.type] ? _errorCounts[data.type] += 1 : _errorCounts[data.type] = 1;

            // implement simple error handling based on hls.js documentation (https://github.com/dailymotion/hls.js/blob/master/API.md#fifth-step-error-handling)
            if (data.fatal) {
                switch (data.type) {
                    case Hls.ErrorTypes.NETWORK_ERROR:
                        console.info("bubbling network error up to VIDEOJS");
                        error.code = 2;
                        tech.error = function() { return error; };
                        tech.trigger('error');
                        break;

                    case Hls.ErrorTypes.MEDIA_ERROR:
                        _handleMediaError();
                        break;

                    default:
                        // cannot recover
                        _hls.destroy();
                        console.info("bubbling error up to VIDEOJS");
                        tech.error = function() { return error; };
                        tech.trigger('error');
                        break;
                }
            }
        }

        function _onMetaData(event, data) {
            var cleanTracklist = [];

            if (data.levels.length > 1) {
                var autoLevel = {
                    id: -1,
                    label: "auto",
                    selected: -1 === _hls.manualLevel
                };
                cleanTracklist.push(autoLevel);
            }

            data.levels.forEach(function(level, index) {
                var quality = {}; // Don't write in level (shared reference with Hls.js)
                quality.id = index;
                quality.selected = index === _hls.manualLevel;
                quality.label = _levelLabel(level);

                cleanTracklist.push(quality);
            });

            var payload = {
                qualityData: {video: cleanTracklist},
                qualitySwitchCallback: switchQuality
            };

            tech.trigger('loadedqualitydata', payload);

            function _levelLabel(level) {
                if (level.height) return level.height + "p";
                else if (level.width) return Math.round(level.width * 9 / 16) + "p";
                else if (level.bitrate) return (level.bitrate / 1000) + "kbps";
                else return 0;
            }
        }

        function _startLoad() {
            _hls.startLoad(startPosition=-1);
            _video.removeEventListener('play', _startLoad);
        }

        function _initHlsjs() {
            // Use underscore to copy hlsjsConfig. hls.js will add defaults in place in this object, and they still be there if we load a second video. This can cause bugs because of invalid config failfasts
            var hlsjsConfig = assign({}, tech.options_.hlsjsConfig);
            var p2pConfig = tech.options_.p2pConfig || {};

            if (['', 'auto'].indexOf(_video.preload) === -1 && !_video.autoplay && hlsjsConfig.autoStartLoad === undefined) {
                hlsjsConfig.autoStartLoad = false;
            }

            // If the user explicitely sets autoStartLoad to false, we're not going to enter the if block above, that's why we have a separate if block here to set the 'play' listener
            if (hlsjsConfig.autoStartLoad === false) {
                _video.addEventListener('play', _startLoad);
            }

            if (!p2pConfig.streamrootKey) {
                p2pConfig = tech.streamrootOptions.p2pConfig;
                hlsjsConfig = tech.streamrootOptions.hlsjsConfig;
            }

            if (!p2pConfig.streamrootKey) {
                throw new Error('No Streamroot key in P2P config');
            }

            _hls = new Hls(hlsjsConfig, p2pConfig);

            _hls.on(Hls.Events.ERROR, function(event, data) { _onError(event, data, tech, _errorCounts) });
            _hls.on(Hls.Events.MANIFEST_PARSED, _onMetaData);
            _hls.on(Hls.Events.LEVEL_LOADED, function(event, data) { _duration = data.details.live ? Infinity : data.details.totalduration; });

            _hls.attachMedia(_video);
            _hls.loadSource(source.src);
        }

        initialize();
    }

    if (Hls.isSupported()) {
        videojs.getComponent('Html5').registerSourceHandler({

            canHandleSource: function (source) {

                var hlsTypeRE = /^application\/x-mpegURL$/i;
                var hlsExtRE = /\.m3u8/i;
                var result;

                if (hlsTypeRE.test(source.type)) {
                    result = 'probably';
                } else if (hlsExtRE.test(source.src)) {
                    result = 'maybe';
                } else {
                    result = '';
                }

                return result;
            },

            handleSource: function (source, tech) {

                if (tech.hlsProvider) {
                    tech.hlsProvider.dispose();
                }

                tech.hlsProvider = new StreamrootProviderHLS(source, tech);

                return tech.hlsProvider;
            }

        }, 0);

    } else {
        console.error("Hls.js is not supported in this browser!");
    }

    videojs.StreamrootProviderHLS = StreamrootProviderHLS;
};

module.exports = attachVideojsStreamrootProvider;
